hips/hip-00NN.md

```diff
@@ -0,0 +1,101 @@
+
+- A Helm chart repository is effectively an `index.yaml` with links to downloads. Maintaining such a repository does create a burden of scripts and automation (e.g. Github Actions). This is not always feasible for smaller projects. It also does not really offer an obvious and readily available way of testing pre-releases.
+- For the testing use cases, charts can be packaged using `helm package`, however this does introduce manual steps and requires extra work to replicate in CI/CD scenarios.
+- There are several plugins available to solve this problem ([`aslafy-z/helm-git`](https://github.com/aslafy-z/helm-git), [`diwakar-s-maurya/helm-git`](https://github.com/diwakar-s-maurya/helm-git), [`sagansystems/helm-github`](https://github.com/sagansystems/helm-github)), however using plugins requires additional setup, which may not always be feasible (esp. in more complex team structures and cluster setups with advanced tooling, e.g. ArgoCD). An additional drawback is that the `Chart.yaml` does not provide a way to specify the plugin requirements, which leaves it up to the consumer to figure this out.
+
+## Rationale
+
+Installing dependencies from git is an established pattern in other ecosystems even when they are registry-based, e.g. npm (Node.js), pipenv (Python), bundler (Gem) have this option - it would make sense to have the behavior replicated in Helm.
+
+At least the npm and the pip ecosystems have already established a syntax of `vcs+protocol` for defining the dependency source, so it should be familiar to some users.
+
+One alternative to consider would be to exclude defining the vcs/protocol, similar to what Go does, esp. given that Helm is built using Go. This does limit the flexibility somewhat - while Go does allow adding a VCS qualifier at the end of the URL (allowing future support for other VCSs), it does not allow specifying the protocol, which means that the users might have to override the default protocol in their VCS configuration.
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
+
+```
+dependencies:
+- name: "<dependency name>"
+  repository: "<protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>"
+  version: "<commit-ish>"
+```
+where:
+- `<protocol>` is one of `git`, `git+ssh`, `git+http`, `git+https`, or `git+file`.
+- `<commit-ish>` is an existing reference (SHA hash, tag or branch name) on the repo.
+
+For example:
+
+```
+dependencies:
+- name: "jenkins"
+  repository: "git+https://github.com/jenkinsci/helm-charts.git/charts/jenkins"
+  version: "main"
+```
+
+When Helm is installing a dependency from git, it should:
+
+- create a temporary directory
+- clone the repo at the specified branch/tag into the temp dir
+- treat the cloned git repo similar to a `file:///path/to/temp/dir` style requirement; use `chart.LoadDir` to load that directory (which in turn applied the logic for filtering the files through `.helmignore`) and archives it to `charts/`
```

Comment:
We should double check the packaging behavior will work correctly here. Helm in places assumes (iirc) that charts with the same (semver) version, represent the same artifact. Probably charts fetched from a git branch will have changes, but retain the same semver version. And this assumption would be invalid.

Reply:
Do you have any clues for specifics? Happy to go look, but the code based has the string "version" in a lot of places and quick scan did not bring results 😅

I can see it maybe might make a difference for deployment/rollback, but that's probably not applicable when used as a dependency?

It would also have to include the registry for comparison, would it not? There's also the checksum in the lock file? And probably the same issue would apply to `file://` dependencies already?

Reply:
Let me investigate, and update with some links.